docs/architecture/adr-071-crypto-v2-multi-curve.md

```diff
@@ -0,0 +1,804 @@
+
+<b>version</b>:
+Indicates the version of the crypto provider using semantic versioning.
+
+<b>configuration (map<string, bytes>)</b>:
+Contains serialized configuration data as key-value pairs, where the key is a string and the value is a byte array.
+
+<b>privKey (google.protobuf.Any)</b>:
+An optional field that can store a private key if it is managed locally.
+
+
+The [record.proto](https://github.com/cosmos/cosmos-sdk/blob/main/proto/cosmos/crypto/keyring/v1/record.proto) file will be modified to include the `CryptoProvider` message as an optional field as follows.
+
+```protobuf
+
+// record.proto
+
+message Record {
+  string name = 1;
+  google.protobuf.Any pub_key = 2;
+
+  oneof item {
+    Local local = 3;
+    Ledger ledger = 4;
+    Multi multi = 5;
+    Offline offline = 6;
+    CryptoProvider crypto_provider = 7; // <- New
+  }
+
+  message Local {
+    google.protobuf.Any priv_key = 1;
+  }
+
+  message Ledger {
+    hd.v1.BIP44Params path = 1;
+  }
+
+  message Multi {}
+
+  message Offline {}
+}
+```
+
+##### Creating and loading a `CryptoProvider`
+
+For creating providers, we propose a *factory pattern* and a *registry* for these builders.
+
+Below, we present the proposed interfaces and code snippets to illustrate the proposed architecture.
+
+```go
+// CryptoProviderFactory is a factory interface for creating CryptoProviders.
+// Must be implemented by each crypto provider.
+type CryptoProviderFactory interface {
+    CreateFromRecord(*Record) (CryptoProvider, error)
+    CreateFromConfig(ProviderMetadata) (CryptoProvider, error)
+}
+```
+
+**Note**: The mechanisms for storing and loading `records` will utilize the existing infrastructure from **Keyring/v1**.
+
+**Code snippet**: provider **factory** and builder **registry**
+
+```go
+// crypto/v2/providerFactory.go
+
+var providerFactories map[string]CryptoProviderFactory
+
+// RegisterCryptoProviderFactory is a function that registers a CryptoProviderFactory for a CryptoProvider.
+func RegisterCryptoProviderFactory(provider CryptoProvider, factory CryptoProviderFactory) {
+    metadata := provider.Metadata()
+    id := generateProviderID(metadata)
+    providerFactories[id] = factory
+}
+
+// CreateCryptoProviderFromRecordOrConfig creates a CryptoProvider based on the provided Record or ProviderMetadata.
+// It enforces that exactly one of Record or ProviderMetadata must be provided.
+func CreateCryptoProviderFromRecordOrConfig(id string, record *Record, config *ProviderMetadata) (CryptoProvider, error) {
+    factory, exists := providerFactories[id]
+    if !exists {
+        return nil, fmt.Errorf("no factory registered for id %s", id)
+    }
+
+    // Validate input parameters
+    if record == nil && config == nil {
+        return nil, fmt.Errorf("both record and config cannot be nil")
+    }
+    if record != nil && config != nil {
+        return nil, fmt.Errorf("both record and config cannot be provided simultaneously")
+    }
```

Comment:
Wouldn't it be simpler, in term of validation, but also UX, to have two different methods, one for creating from record, and the other for creating from config?

Reply:
Sure, it can be done that way. This code is just an example and may not be in the final implementation.